[1단계 - 로또(자동)] 알린(장원영) 미션 제출합니다.

OzRagwort opened the pull request:
안녕하세요 닉! 알린입니다

이번 미션 잘 부탁드립니다!

wyRagwort committed:
docs(README): 기능 요구 사항 작성

wyRagwort committed:
feat(lotto): 구입 금액 입력 기능 구현

wyRagwort committed:
feat(lotto): 로또 구매 기능 구현

wyRagwort committed:
fix(LottoGenerator): 로또 생성기 로직 수정

wyRagwort committed:
feat(lotto): 구매한 로또 정보 출력 기능 구현

wyRagwort committed:
feat(LottoNumber): LottoNumber VO 추가

wyRagwort committed:
feat(lotto): 당첨 번호, 보너스 볼 입력 기능 구현

wyRagwort committed:
feat(lotto): 당첨 통계 출력 기능 구현

wyRagwort committed:
feat(lotto): 수익률 계산 및 출력 기능 구현

wyRagwort committed:
fix(LottoPrize): 로또 당첨 등수 확인 기능 오류 수정

wyRagwort committed:
refactor(LottoGenerator): 로또 번호 생성 알고리즘 수정

wyRagwort committed:
refactor: 도메인에서 unmodifiable로 반환 값이 불변하도록 함

wyRagwort committed:
feat(LottoResults): 당첨 결과에서 등수가 누락된 경우를 검증하는 기능 구현

wyRagwort committed:
refactor: vo를 vo 패키지로 이동

wyRagwort committed:
refactor: LottoGenerator가 당첨 번호가 아닌 로또를 반환하도록 수정

wyRagwort committed:
fix(test): 테스트가 정상적으로 동작하도록 수정

OzRagwort commented on src/test/java/lotto/domain/LottosTest.java:
```diff
@@ -19,6 +19,6 @@ void purchase_test() {
         lottos.purchase(new Money(10000));
 
-        assertThat(lottos.matchSize(10)).isTrue();
+        assertThat(lottos.getLottos().size()).isEqualTo(10);
```
테스트에서 getter를 이용하여 테스트하는 것에 대한 닉의 의견이 궁금합니다. 

테스트를 위해서 getter를 만들지는 않겠지만 다른 목적으로 만든 getter가 있다면 테스트에서 getter를 사용해도 괜찮을까요?

wbluke replied:
보통 getter 같은 경우는 데이터가 필요한 시점에 사용할 수밖에 없기 때문에 테스트에서도 getter를 사용하는 것이 일반적입니다.
내부 데이터가 잘 생성이 되었는지 검증하려면 어쩔 수 없겠죠 ㅎㅎ
그리고 말씀하신 것처럼 지금 당장 사용하지 않더라도 위와 같은 이유로 (미래에 사용할 가능성이 높기 때문에) 테스트를 위해서 getter를 열어놔도 괜찮다고 생각해요 :)

OzRagwort replied:
아하 감사합니다 :)

wbluke commented on src/main/java/lotto/view/ResultView.java:
```diff
@@ -0,0 +1,58 @@
+    private static final String LOTTO_RESULT_DELIMITER = "---------";
+    private static final String LOTTO_RESULT_TWICE_MESSAGE = "%d개, 보너스 볼 일치 일치 (%d원)- %d개" + System.lineSeparator();
+    private static final String LOTTO_RESULT_DEFAULT_MESSAGE = "%d개 일치 (%d원)- %d개" + System.lineSeparator();
+    private static final String LOTTO_RATE_RETURN_MESSAGE = "총 수익률은 %.2f입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)";
```
`(기준이 1이기 때문에 결과적으로 손해라는 의미임)`
이 멘트는 요구사항 설명을 위한 문구라 빼는 것이 좋겠어요.
사용자가 수익을 냈는데 위 문구를 보면 당황하지 않을까요? ㅎㅎ 
클라이언트 사이드에서 고려하는 것도 굉장히 중요한 포인트 중 하나입니다 :)

OzRagwort replied:
생각해보니 수익률이 1 이상인 경우 사용자 입장에서 수익률이 잘못된건지 문구가 잘못된건지도 헷갈릴 것 같네요.

클라이언트 사이드에서 고려하는 것도 굉장히 중요하다 꼭 기억하겠습니다!

wbluke commented on src/main/java/lotto/domain/vo/Money.java:
```diff
@@ -0,0 +1,27 @@
+
+    private static final String ERROR_NEGATIVE_INPUT_MESSAGE = "금액은 음수를 입력하면 안됩니다.";
+
+    private int money;
```
VO라면 다음 사항을 항상 고려해 주세요 :)
- 필드가 변하지 않을 값이기 때문에 final
- eq & hc 재정의

OzRagwort replied:
이전 미션( [자동차 미션을 하며 생각해본 VO](https://github.com/woowacourse/java-racingcar/pull/387#issuecomment-1046303133) )을 하면서 VO에서 `equals`와 `hashCode`를 재정의하고 가능하면 `toString()`까지 재정의해주면 좋다는 것을 확인하고도 의식하지 못한것같습니다. 

다시한번 간단하게 정리하고 다음부터는 꼭 기억하도록 노력하겠습니다!

- 왜 불변해야하는가? 의도하지 않은 수정(사이드 이펙트)을 막기 위해
- 왜 equals, hashCode를 재정의 해야하는가? 동등성 검사를 위해(이번 미션을 하다보니 contains 같은것들도 equals를 재정의하지 않으면 제대로 동작하지 않는 것 같음)
- 왜 validate를 해야하는가? VO가 가진 값을 VO 내부에서 검증하여 응집도가 높아진다.

wbluke commented on src/main/java/lotto/domain/vo/Money.java:
```diff
@@ -0,0 +1,27 @@
+
+    private int money;
+
+    public Money(int money) {
+        validateNegative(money);
+        this.money = money;
+    }
```
현재 애플리케이션에서 구입금액 500원을 입력하면 수익률이 NaN이 나와요.
방법은 두 가지 일 것 같아요.
1. 1000 미만의 금액이 들어오면 예외 발생
2. 1000 미만의 금액이 들어오면 당첨번호를 입력 받기 전에 애플리케이션을 종료시키거나, 수익률 오류가 나지 않도록 수정

만약 1번을 선택한다면 Money 에 대한 네이밍도 더 구체적으로 해야할 것 같아요 :)

wbluke replied:
다시보니까 상금에도 쓰이고 있는 객체였네요 ㅎㅎ
유연한 방향으로 고민해 보셔요 :)

OzRagwort replied:
이 문제는 로또를 0개 구매하면 수익률 계산 시 0으로 나누는 식이 생겨 발생했습니다. 해결하기 위해서는 말씀해주신 세 가지 방법이 있을 것 같습니다. 

돈이 없어서 로또를 사지 못하는 상황에서 세 가지 방법으로 생각을 해봤습니다.

- 돈이 부족하여 로또를 사지 못하면 (예외 발생 <-> 0개 구매 후 진행)
- 로또를 0개 가지고 있을 때 (당첨 번호 입력 <-> 미입력)
- 로또를 0개 가지고 있을 때 (당첨 통계 출력 <-> 미출력)

저는 돈이 부족하여도 0개를 구매한 후 당첨 번호를 입력하지 않고 당첨 통계도 출력하지 않는 방향으로 수정했습니다.

돈이 부족한 것은 입력에서 금액을 음수/문자로 입력하는 것과 같이 예외를 발생시킬 문제는 아니라고 생각했습니다. 그래서 0개를 구매하도록 했습니다. 하지만 로또의 수가 0개라면 당첨이 될 수 없기 때문에 당첨 번호를 입력받거나 결과를 출력하지 않고 프로그램을 종료했습니다.

---
상금에서도 Money를 사용하고 있는데 적용해보지는 않았지만 최근에 배운 인터페이스를 적용해볼까도 고민해봤습니다.

Money를 인터페이스로 만들고 구입금액 구현체와 상금 구현체를 만들어보면 구입 금액에서는 1000원 이상만 받도록 검증을 할 수 있을 것 같았기 때문입니다. 하지만 아직 인터페이스가 많이 헷갈리기도 하고 예외발생을 안시키는 방향으로 수정하다보니 실제로 적용해보지는 않았던것 같습니다. 

이런 방법으로 구현하는 것도 인터페이스를 올바르게 사용하는 걸까요? 아직 인터페이스를 제대로 사용해본적이 거의 없는 것 같아서 어렵네요..!

wbluke replied:
네네 말씀하신 방식으로 구현해볼 수도 있겠네요 ㅎㅎ
인터페이스는 앞으로도 사용해볼 기회가 많으니 차차 친숙해지는걸로 해요 :)

wbluke commented on src/main/java/lotto/domain/vo/generator/LottoGenerator.java:
```diff
@@ -0,0 +1,33 @@
+    private static final int LOTTO_MINIMUM = 1;
+    private static final int LOTTO_MAXIMUM = 45;
+    private static final int FROM_LOTTO_INDEX = 0;
+    private static final int TO_LOTTO_INDEX = 6;
```
- 1, 45는 LottoNumber의 상수 아닐까요?
- 6이라는 숫자는 Lotto의 상수 아닐까요?
인덱스 0은 상수로 추출 안해도 될 것 같고, 다른 객체의 상수를 이용하도록 해보면 좋을 것 같아요 :)

OzRagwort replied:
이미 다른곳에 있는 상수인건 알고있었습니다. 그런데 private로 상수가 선언되어있어서 가져오려면 getter를 써야하나 고민하다가 새로 정의를 해버린것같습니다. 

이 문제를 해결하기 위해 고민하던중 Integer의 MAX_VALUE와 MIN_VALUE처럼 public 으로 상수를 만들어 보기로 하고 그 상수를 가져와서 사용했습니다. 

프리코스때부터 무의식적으로 상수는 무조건 private로 만들어야한다고 생각하고 있었던 것 같습니다. 

어떤 상수 값이 필요한 경우 상수를 public으로 만들어 사용하는 것도 괜찮은 방법인가요? 그래도 private로 만들고 getter를 사용하는 방법이 더 좋을까요?

wbluke replied:
상수는 말그대로 변하지 않는 값이기 때문에, 굳이 getter를 통할 필요 없이 public으로 열어도 된다고 생각해요 :)

OzRagwort replied:
맞는것 같아요 감사합니다! 😃

wbluke commented on src/main/java/lotto/domain/vo/generator/LottoGenerator.java:
```diff
@@ -0,0 +1,33 @@
+package lotto.domain.vo.generator;
```
패키지가 조금 이상한데요. 알린이 생각하는 VO의 정의가 무엇인가요?
VO 패키지 안에 VO인 것과 VO가 아닌 것이 섞여있어요.
한번 다시 배치해보면 좋겠네요 :)

OzRagwort replied:
generator는 VO가 아니지만 VO를 만들어 주기 때문에 VO 패키지에 넣어도 되지 않나 하고 쉽게 생각했던것 같습니다. 

VO는 원시 값, 문자열, 다른 객체 등을 멤버로 가지고 있는 객체라고 생각합니다. 특징으로는 값이 불변성해야 하고, VO와 VO간의 동등성을 비교할 수 있어야 하고, 값을 역할에 따라 검증할 수 있어야 한다고 이해했습니다. 

이해는 했지만 아직도 헷갈리는 것 같습니다. 특히 일급 컬렉션이 그렇습니다. 

[자동차 미션을 하며 생각해본 VO](https://github.com/woowacourse/java-racingcar/pull/387#issuecomment-1046303133) 에서 저는 일급 컬렉션은 final로 재할당을 방지하고 어떤 반환이 필요할 경우 `Collections.unmodifiableList()`를 쓰면 될 것 같다고 했지만 누군가가 고의든 실수든 값을 변경하는 메소드를 만든다면 불변성을 깨지게 되는 문제가 있습니다. 

이번 미션에서도 Lottos의 `purchase()`는 Lottos의 `List<Lotto>`에 add를 하여 값을 변경시키는 메서드였습니다. 나름대로 신경썼고 제가 짠 코드였지만 지적해주시지 않았다면 계속 모르고 있었을 것 같습니다. 

앞으로도 익숙해질때까지 많이 연습해야할 것 같네요! 😢

wbluke commented on src/main/java/lotto/domain/vo/WinningNumbers.java:
```diff
@@ -0,0 +1,56 @@
+
+    public WinningNumbers(List<LottoNumber> lottoNumbers, LottoNumber bonusNumber) {
+        validateNumberSix(lottoNumbers);
+        validateLottoNumbersDuplication(lottoNumbers);
```
Lotto에서 이미 체크하고 있는 validation 아닌가요?
`List<LottoNumber>`를 이미 존재하는 일급컬렉션을 사용하도록 하면 좋을 것 같아요 :)

OzRagwort replied:
같은 validation이 맞는데 저부분을 신경쓰지 못했네요. 

`List<LottoNumber>`를 Lotto로 수정하여 Lotto가 검증하도록 수정했습니다!

wbluke commented on src/main/java/lotto/domain/LottoPrize.java:
```diff
@@ -0,0 +1,53 @@
+        }
+        if (prize.equals(LottoPrize.THIRD)) {
+            return lottoNumberMatches == prize.lottoNumberMatches && bonusNumberMatches == prize.bonusNumberMatches;
+        }
```
2등 3등 분기를 하나로 합쳐도 될 것 같은데요? :)

OzRagwort replied:
아직 stream을 다루는게 어색하다보니까 불필요하게 조건문이 길어진 것 같습니다. 

저 코드는 [46ebb2f](https://github.com/woowacourse/java-lotto/pull/374/commits/46ebb2f396fd0c98c9615f45e6b5a42c191d9dda) 처럼 수정했습니다.

그리고 수정을 하고보니 LottoPrize enum에서 bonusNumber를 맞추는지 아닌지를 알 필요가 없어져서 LottoPrize enum도 함께 수정되었습니다.

wbluke commented on src/main/java/lotto/domain/vo/Lottos.java:
```diff
@@ -0,0 +1,41 @@
+        for (int i = 0; i < purchaseNumber; i++) {
+            lottos.add(LottoGenerator.generate());
+        }
+    }
```
Lotto 여러 개를 묶은 Lottos 일급컬렉션에서 `purchase()`라는 구매 행위를 하는 것이 조금 어색해 보여요.
제 3자가 구매를 해서 발생한 결과물이 Lottos여야 할 것 같아요.
LottoGame 객체도 만드셨으니 한번 구조 변경을 해서 객체의 역할과 책임을 재설정해보면 좋겠네요 :)

OzRagwort replied:
처음에는 구매 행위를 하는 것이 어색하다는 의미를 잘 이해하지 못했습니다.

그런데 VO의 특징을 생각해보던 중 Lottos의 `purchase()`는 Lottos의 `List<Lotto>`에 add를 하여 값을 변경시키니 가변이고, 그래서 어색하다고 하셨던건가 싶었습니다. 그래서 LottoGame에서 로또를 만들고 만들어진 Lotto들을 Lottos로 만들도록 수정하였습니다. 

수정하면서 로또 생성의 책임이 LottoGame으로 옮겨지고 Lottos가 가지고 있던 로또 1개의 가격은 LottoGame이 가지게 되었습니다.

wbluke commented on src/test/java/lotto/domain/LottoResultsTest.java:
```diff
@@ -0,0 +1,79 @@
+                .withMessage("로또 결과가 잘못되었습니다.");
+    }
+
+    @DisplayName("getRateReturn 메소드 테스트")
```
테스트를 작성할 때 고려할 점은, 이런 테스트 하나의 명세를 정할 때 되도록 추상적인 표현을 뺀 구체적인 문장으로 표현해보세요.
테스트의 역할에서 가장 중요한 것 중 하나는 **프로덕션 코드에 대한 명세(문서)** 라는 것입니다.
테스트만 보고서 프로덕션 코드가 어떻게 동작하겠구나, 어떤 기능이 있구나 하는 것을 알 수 있어야 합니다.

`getRateReturn 메소드 테스트` 라는 표현을 봐서는 어떤 테스트 내용인지 전혀 알수가 없겠죠. ㅎㅎ
`당첨결과를 계산하여 수익률을 반환한다` 라는 표현이 훨씬 와닿을 거에요.

한번 전체적으로 체크해보셔요 :)

OzRagwort replied:
단위 테스트가 메서드를 설명하는 문서가 된다는 것을 명심하고 DisplayName을 더 구체적으로 작성할 수 있도록 노력해야겠습니다. 

내가 짠 테스트와 코드는 내가 작성했기 때문에 DisplayName을 추상적으로 적어도 어느정도는 괜찮겠지만 다른 개발자와 시간이 지난 뒤 내가 다시 볼 때 추상적인 DisplayName은 코드 이해가 늦어지는 원인이 될 것 같습니다. 

하지만 너무 길어도 안될 것 같아서 적당한 길이이지만 최대한 구체적으로 DisplayName을 [수정](https://github.com/woowacourse/java-lotto/pull/374/commits/caef316d9ae62627048cce65ba80acbc3e195253) 해봤습니다.

wbluke commented on src/test/java/lotto/domain/vo/LottosTest.java:
```diff
@@ -0,0 +1,71 @@
+        Map<LottoPrize, Integer> testMap = initTestMap(lottos, winningNumbers);
+        Map<LottoPrize, Integer> lottoMatches = lottos.confirmWinnings(winningNumbers);
+
+        assertThat(testMap).isEqualTo(lottoMatches);
```
이런 검증 방식을 위해서 아래에 private 메서드가 너무 과하게 생겨난 것 같아요.
Map 두개를 비교하기 보다, confirmWinnings()의 결과물인 Map을 가지고
값을 하나씩 꺼내서 내가 예상한 값과 같은지를 보는 것이 더 올바른 테스트라고 생각해요 :)

OzRagwort replied:
[수정 커밋](https://github.com/woowacourse/java-lotto/pull/374/commits/a6a481155153968a1ebc445462804f63a58510f5)

아직 조금 어렵지만 전략 패턴으로 인터페이스를 분리해 테스트를 위한 CustomLottoGenerator를 만들어 사용했습니다. 

CustomLottoGenerator는 `generate()`에서 고정적으로 로또 번호가 1,2,3,4,5,6이 생성되는 Generator입니다. 따라서 당첨 결과를 예측할 수 있고 불필요한 과정을 제거하여 테스트를 간단하게 바꿀 수 있었습니다.

wbluke reviewed (changes requested):
안녕하세요 알린~! 로또 미션 리뷰를 맡은 닉입니다 🙌 
일급컬렉션을 사용하여 잘 구현해주셨네요 ㅎㅎ
좀 더 개선할 수 있는 포인트 남겼으니 확인해주시고 수정 부탁드려요 🙂

wyRagwort committed:
docs(README): 피드백 및 수정할 것 정리

wyRagwort committed:
refactor(ResultView): 수익률에 따라 손해, 이득, 본전인지 출력하도록 수정

wyRagwort committed:
refactor: 구입금액으로 로또를 사지 못할 때 당첨 번호 입력, 당첨 통계 출력을 하지 않도록 수정

wyRagwort committed:
refactor(LottoGenerator): 패키지 위치를 VO 외부로 이동

wyRagwort committed:
refactor(VO): vo에 equals, hashCode, toString 재정의

wyRagwort committed:
refactor(Lottos): LottosGame에서 구매하여 Lottos를 생성하도록 수정

wyRagwort committed:
refactor(LottoGenerator): 다른 클래스의 상수를 사용하도록 수정

- LottoNumber : LOTTO_MAXIMUM, LOTTO_MINIMUM을 public으로 변경
- Lotto : LOTTO_SIZE를 public으로 수정

wyRagwort committed:
refactor(WinningNumber): Lotto를 이용하여 검증 책임을 분리

wyRagwort committed:
refactor(LottoPrize): 당첨 결과를 반환하는 메서드 stream 수정

wyRagwort committed:
refactor: LottoGame.purchase 테스트를 위해 Generator 인터페이스 분리

wyRagwort committed:
refactor(LottosTest): 필수 검증만을 하여 불필요한 private 메서드 제거

wyRagwort committed:
refactor(test): DisplayName을 구체적으로 작성

OzRagwort commented:
안녕하세요 닉! 알린입니다

주신 피드백을 보고 수정하면서 고민해본것이나 궁금한 것은 댓글로 남겨봤습니다. 

이해하기 쉽게 잘 피드백 해주셔서 감사합니다!!

wbluke commented on src/main/java/lotto/domain/LottoPrize.java:
```diff
@@ -5,42 +5,42 @@
         return Arrays.stream(LottoPrize.values())
-                .filter(prize -> checkMatches(prize, lottoNumberMatches, bonusNumberMatches))
+                .filter(prize -> prize.lottoNumberMatchCount == lottoNumberMatchCount)
+                .filter(prize -> !matchTwiceOrThird(prize) || judgeTwiceAndThird(prize, bonusNumberMatch))
```
요 부분을 한 눈에 이해하기가 어려운 것 같아요.
더 나은 방향으로 고민해보면 어떨까요?
(ex. 2등을 먼저 골라서 early return 하는 방법)

OzRagwort replied:
이 부분에서 자꾸 2등과 3등을 제대로 비교하지 못해 Enum 순서가 바뀌거나 하면 제대로 값이 안나오는 문제가 있었습니다. 

그래서 몸을 꼬아가며 작성하다보니 이해하기는 더 어려운 코드가 되어버린것같네요

생각해보면 보너스 번호에 영향을 받는 2등만 early return으로 처리하면 너무 해결하기 쉬운 문제였던것 같습니다.

제가 문제가 생기면 자꾸 어렵게만 생각하는 것 같은데 쉽게쉽게 생각해야겠습니다 😃

wbluke commented on src/test/java/lotto/domain/vo/LottosTest.java:
```diff
@@ -3,69 +3,57 @@
     void confirmWinnings_test() {
-        Lottos lottos = new Lottos();
-        lottos.purchase(new Money(10000));
+        Generator lottoGenerator = new CustomLottoGenerator();
+        Lottos lottos = new Lottos(new ArrayList<>() {{
+            add(lottoGenerator.generate());
+        }});
```
음 저는 조금 의견이 다른데요 ㅎㅎ
CustomLottoGenerator를 만드는 시도는 좋았어요 👍 
다만 테스트는, 테스트하고자 하는 내용을 한눈에 알아볼 수 있어야 하는데요.
해당 커스텀 Generator를 들어가서 내용을 확인해야 위 테스트의 내용을 파악할 수 있게 됩니다.
LottoNumber 리스트를 단순히 생성하는 것이 더 좋을 것 같아요 :)

OzRagwort replied:
테스트가 명세가 된다는 것과 테스트하려는 것을 알기 쉬워야 한다는 것을 생각하다 보니 제 코드에서 문제가 참 많다고 느꼈습니다.

그래서 2단계 미션을 하면서 테스트 코드를 어떻게 더 쉽고 정확하게 할 수 있을까를 생각해봤던것 같습니다!

wbluke commented on src/main/java/lotto/domain/vo/LottoNumber.java:
```diff
@@ -6,8 +6,8 @@
 
```
LottoNumber에 대해 한번 더 고민하고 발전시켜볼까요?
현재는 45개만 존재하는 LottoNumber를 같은 숫자여도 매번 생성할 수 있게 되는데요.
동일한 1번 LottoNumber라면 항상 같은 인스턴스를 사용할 수 있도록 캐싱을 해보면 어떨까요?

힌트)
- 캐싱할 자료구조를 고민한다.
- 어떤 숫자로 LottoNumber 생성 요청이 왔을 때, 해당 자료구조에 기존에 만들어둔 인스턴스가 있는지 확인하고, 있으면 그대로 반환, 없으면 새로 생성한 후 캐싱한 뒤 반환한다.
- 캐싱한 자료구조를 위한 별도 클래스 (pool)를 만들어도 좋다. (혹은 LottoNumber 내부에 있어도 좋다.)
- 위 경로 외에 LottoNumber를 생성할 수 있는 다른 방법이 있어서는 안 된다. (캐싱하는 의미가 없어지므로)

OzRagwort replied:
정적 팩터리 메서드와 플라이웨이트 패턴으로 LottoNumber를 캐싱해봤습니다.

사실 제가 한 부분이 플라이웨이트가 맞는지 캐싱이 되긴 하는데 다른 문제는 없는지 아직은 조금 자신이 없지만 캐싱으로 LottoNumber를 수정하면서 많이 신기하다고 느꼈습니다.

많이 고민할 수 있는 주제를 주셔서 감사합니다! 😄

wbluke reviewed (approved):
피드백 반영 잘 해주셨네요 👍 
로또 숫자에 대한 고민을 좀 더 발전시켜 보았어요. 한번 같이 고민하시면서 2단계 진행 부탁드려요 🙂
